### PR TITLE
Hooked up secrets function api with secret function implementations

### PR DESCRIPTION
**Pull request by srkukarni**
### Motivation

This pr defines the secrets interface for java and python functions and hooks their implementations.

### Modifications

Describe the modifications you've done.

### Result

After your change, what will change.

### CODE REVIEW

**Comment by srkukarni** [comment]
@cckellogg

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java [hunk 1]
```diff
@@ -49,6 +50,8 @@
                                            Integer grpcPort,
                                            Long expectedHealthCheckInterval,
                                            String logConfigFile,
+                                           String secretsProviderClassName,
```
Comment (cckellogg) [inline pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: Can some docs be added to what the secretsProviderConfig could or does contain. Should the classname just go in the secretsProviderConfig  so there is only one variable?

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/ThreadRuntimeFactory.java [hunk 2]
```diff
@@ -44,15 +53,25 @@
...
+                                Map<String, String> secretsProviderConfig) throws Exception {
+        this(threadGroupName, createPulsarClient(pulsarServiceUrl, authConfig), storageServiceUrl, secretsProviderClassName, secretsProviderConfig);
     }
 
     @VisibleForTesting
-    public ThreadRuntimeFactory(String threadGroupName, PulsarClient pulsarClient, String storageServiceUrl) {
+    public ThreadRuntimeFactory(String threadGroupName, PulsarClient pulsarClient, String storageServiceUrl,
+                                String secretsProviderClassName, Map<String, String> secretsProviderConfig) {
+        try {
+            secretsProvider = (SecretsProvider) Reflections.createInstance(secretsProviderClassName, ClassLoader.getSystemClassLoader());
```
Comment (cckellogg) [inline pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/ThreadRuntimeFactory.java]: I don't think this should be enforced it's an optional thing and no implementation should be the default. Also just enable this for kubernetes runtime to start then add the others in additional patches.

File: pulsar-functions/worker/src/test/java/org/apache/pulsar/functions/worker/SchedulerManagerTest.java [hunk 3]
```diff
@@ -783,7 +784,7 @@ public void testAssignmentWorkerDoesNotExist() throws InterruptedException, NoSu
         functionMetaDataList.add(function2);
         doReturn(functionMetaDataList).when(functionMetaDataManager).getAllFunctionMetaData();
 
-        ThreadRuntimeFactory factory = new ThreadRuntimeFactory("dummy", null, "dummy");
+        ThreadRuntimeFactory factory = new ThreadRuntimeFactory("dummy", null, "dummy", ClearTextSecretsProvider.class.getName(), null);
```
Comment (cckellogg) [inline pulsar-functions/worker/src/test/java/org/apache/pulsar/functions/worker/SchedulerManagerTest.java]: can we have some of these test pass something other than null for the config

File: pulsar-functions/runtime/src/test/java/org/apache/pulsar/functions/runtime/ProcessRuntimeTest.java [hunk 4]
```diff
@@ -149,7 +152,8 @@ public void testPythonConstructor() throws Exception {
                 + " --function_details '" + JsonFormat.printer().omittingInsignificantWhitespace().print(config.getFunctionDetails())
                 + "' --pulsar_serviceurl " + pulsarServiceUrl
                 + " --max_buffered_tuples 1024 --port " + args.get(23)
-                + " --expected_healthcheck_interval 30";
+                + " --expected_healthcheck_interval 30"
+                + " --secrets_provider secretsprovider.ClearTextSecretsProvider";
```
Comment (cckellogg) [inline pulsar-functions/runtime/src/test/java/org/apache/pulsar/functions/runtime/ProcessRuntimeTest.java]: we should have a tests passing the secret configs in

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/ThreadRuntimeFactory.java (hunk 2 above)
Comment (cckellogg) [reply pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/ThreadRuntimeFactory.java]: In general it's not a good idea to throw exceptions in a constructor. Is only the secretsProvider object needed in the class? Can the instantiation and configuration happen outside in utility or factory method and then passed in?

Comment (srkukarni) [reply pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/ThreadRuntimeFactory.java]: Done

File: pulsar-functions/worker/src/test/java/org/apache/pulsar/functions/worker/SchedulerManagerTest.java (hunk 3 above)
Comment (srkukarni) [reply pulsar-functions/worker/src/test/java/org/apache/pulsar/functions/worker/SchedulerManagerTest.java]: Changed the signature

File: pulsar-functions/runtime/src/test/java/org/apache/pulsar/functions/runtime/ProcessRuntimeTest.java (hunk 4 above)
Comment (srkukarni) [reply pulsar-functions/runtime/src/test/java/org/apache/pulsar/functions/runtime/ProcessRuntimeTest.java]: Added

File: pulsar-functions/instance/src/main/python/python_instance_main.py [hunk 5]
```diff
@@ -120,11 +123,23 @@ def main():
   if args.tls_trust_cert_path:
      tls_trust_cert_path =  args.tls_trust_cert_path
   pulsar_client = pulsar.Client(args.pulsar_serviceurl, authentication, 30, 1, 1, 50000, None, use_tls, tls_trust_cert_path, tls_allow_insecure_connection)
+
+  secrets_provider = None
```
[the remark this answers is not in the document]
Comment (srkukarni) [reply pulsar-functions/instance/src/main/python/python_instance_main.py]: Added for secrets provider

**Comment by srkukarni** [comment]
rerun integration tests

File: pulsar-functions/instance/src/main/python/secretsprovider.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+"""secretsprovider.py: Interfaces and definitions for Secret Providers
+"""
+from abc import abstractmethod
+import os
+
+class SecretsProvider:
+  """Interface for providing secrets information runtime"""
+  @abstractmethod
+  def init(self, config):
```
Comment (jerrypeng) [inline pulsar-functions/instance/src/main/python/secretsprovider.py]: should we call this init since python constructor is __init__

Comment (srkukarni) [reply pulsar-functions/instance/src/main/python/secretsprovider.py]: Python Constructor is actually __init__

**Pull request merged**